### DISCUSSION TITLE
Upload temporarily files to bucket via S3

### DISCUSSION THREAD

**Discussion by meotimdihia** [seaweedfs/seaweedfs#2824] [Q&A]
Hi, I want to upload files to a bucket then Seaweedfs automatically deletes them later.
It is TTL feature that Seaweedfs had, but I can't get it works.
I read 2 discussions: 

**s3 bucket set TTL** https://github.com/chrislusf/seaweedfs/discussions/2718
[chrislusf](https://github.com/chrislusf) confirmed we can't set TTL for bucket

**S3 gateway and TTL** https://github.com/chrislusf/seaweedfs/discussions/2199
But in this discussion, it seems we can do something about this. But I don't really understand it. 

This is my `filer.conf` - I want to set `ttl 1w` to bucket `temp` 
```json
{
  "locations": [
    {
      "locationPrefix": "/buckets/temp",
      "collection": "temp",
	  "ttl": "1w"
    }
  ]
}
```

P/S: because I configured security for endpoints with JWT. And then now, I don't know to generate tokens for this example: https://github.com/chrislusf/seaweedfs/wiki/Store-file-with-a-Time-To-Live
I checked javascript library, but they don't have this feature too.

**Comment by guo-sj** [comment]
You can't set TTL to an S3 bucket as Seaweedfs doesn't support that.

As written in [wiki/Store-file-with-a-Time-To-Live](https://github.com/chrislusf/seaweedfs/wiki/Store-file-with-a-Time-To-Live), there are only two ways to set TTL:
1. For collection, you can set collections' TTL in `weed shell` by this command:
```
> fs.configure -locationPrefix=/my/folder -collection=abc -ttl=7d -apply
```
Then you can post a file to the path `/my/folder` to activate this collection TTL:
```
$ curl -F file=@filename "http://localhost:8888/my/folder
```

2. For normal files, you can set their TTL by the `curl` command in your OS shell:
```
$ curl http://localhost:9333/dir/assign?ttl=3m   # get a fid
{"count":1,"fid":"5,01637037d6","url":"127.0.0.1:8080","publicUrl":"localhost:8080"}

$ curl -F "file=@x.go" http://127.0.0.1:8080/5,01637037d6?ttl=3m  # sent your file and set its TTL
```

BTW, you can open the log to see its behavior about TTL by adding `v=4` in starting volume server:
```
$ weed -v=4 volume -dir=/tmp/data2 -max=40 -mserver=localhost:9333 -port=8081
```

**Reply by meotimdihia** [reply]
Thanks for the answer. But the problem, I secured endpoints with JWT, now I don't know how to generate tokens.
I tested with commands:

```shell
curl -D - "http://localhost:9333/dir/assign?ttl=3m
```
**output**
HTTP/1.1 200 OK
Authorization: BEARER eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmaWQiOiIxMzAwNCwwMWE0OGMxNTY2YTA1OGMyIiwiZXhwIjoxNjQ4NDMxMTY2fQ.d3yQlTsjKYtfB1Plr5bub-OTd5UEdAgUlituR9EHIcQ
Content-Type: application/json
Date: Mon, 28 Mar 2022 01:32:36 GMT
Content-Length: 104

{"fid":"13004,01a48c1566a058c2","url":"142.132.153.23:8080","publicUrl":"xxxxxxxxxxxxxx:8080","count":1}

```shell
curl -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmaWQiOiIxMzAwNCwwMWE0OGMxNTY2YTA1OGMyIiwiZXhwIjoxNjQ4NDMxMTY2fQ.d3yQlTsjKYtfB1Plr5bub-OTd5UEdAgUlituR9EHIcQ" -F "file=@test.txt" "http://127.0.0.1:8080/13004,01a48c1566a058c2?ttl=3m"
```
**output**
{"error":"wrong jwt"}

**Reply by guo-sj** [reply]
I have no idea about this problem, but I found something about JWT in wiki, maybe they can help you:
1. https://github.com/chrislusf/seaweedfs/wiki/Security-Overview#jwt-based-access-control
2. https://github.com/chrislusf/seaweedfs/wiki/Security-Configuration

**Comment by theRealWardo** [comment]
In #2718 @chrislusf confirmed you cannot set a TTL for a bucket **directory** and in #2199 he confirmed you can set a TTL using `fs.configure` in the `weed shell`.

To set a TTL for an S3 bucket, you can configure the associated collection with a TTL. The collection is named the same as your S3 bucket, so if you are writing to `s3://test-bucket/path` a collection named `test-bucket` will be written to by default. Suppose you want this whole bucket to have a 5 minute TTL:

```
> fs.configure -locationPrefix=/ -collection=test-bucket -ttl=5m -apply
{
  "version":  0,
  "locations":  [
    {
      "locationPrefix":  "/",
      "collection":  "test-bucket",
      "replication":  "",
      "ttl":  "5m",
      "diskType":  "",
      "fsync":  false,
      "volumeGrowthCount":  0,
      "readOnly":  false,
      "dataCenter":  "",
      "rack":  "",
      "dataNode":  ""
    }
  ]
}
```

Upload a file with the AWS CLI - `aws --endpoint-url http://localhost:8333 s3 cp local-test-file s3://test-bucket/test`

To verify the TTL is correctly set for a file:
```
> fs.meta.cat /buckets/test-bucket/test
{
  "name":  "test",
  "isDirectory":  false,
  "chunks":  [
  ...
  ],
  "attributes":  {
    "fileSize":  "29380050",
    "mtime":  "1668454574",
    "fileMode":  504,
    "uid":  0,
    "gid":  0,
    "crtime":  "1668454574",
    "mime":  "",
    "ttlSec":  300,
    "userName":  "",
    "groupName":  [],
    "symlinkTarget":  "",
    "md5":  "",
    "rdev":  0,
    "inode":  "0"
  },
  "extended":  {},
  "hardLinkId":  "",
  "hardLinkCounter":  0,
  "content":  "",
  "remoteEntry":  null,
  "quota":  "0"
}chunks 8 meta size: 609 gzip:509
```

Note the `.attributes.ttlSec` value is correctly set. If you wait the TTL time and run `fs.ls` in `weed shell` you should no longer see the file.
